src/lightning_fabric/utilities/data.py

```diff
@@ -42,31 +43,34 @@ def __call__(self, *args: Any) -> None:
         return fn(*args)
 
 
-def has_iterable_dataset(dataloader: DataLoader) -> bool:
-    return hasattr(dataloader, "dataset") and isinstance(dataloader.dataset, IterableDataset)
+def has_iterable_dataset(dataloader: object) -> TypeGuard[DataLoader]:
+    return isinstance(dataloader, DataLoader) and isinstance(dataloader.dataset, IterableDataset)
 
 
-def has_len(dataloader: Union[DataLoader, Iterable, Dataset]) -> bool:
-    """Checks if a given Dataloader has ``__len__`` method implemented i.e. if it is a finite dataloader or
-    infinite dataloader."""
+def sized_len(dataloader: object) -> Optional[int]:
```

Comment:
nit: instead of `object` I'd prefer `Any`, which should be the same (applies to other parts of this PR as well)

Reply:
object is more informative, as Any signals the type checker to ignore this path entirely: https://stackoverflow.com/a/39817126